src/irmin/data/fixed_size_string_set.ml

```diff
@@ -0,0 +1,237 @@
+  val contains : hashset -> t -> elt -> bool
+
+  val contains_substring :
+    hashset -> t -> src:Bigstringaf.t -> src_off:int -> bool
+
+  val is_empty : hashset -> t -> bool
+  val get : hashset -> t -> elt
+  val set : hashset -> t -> elt -> unit
+  val set_substring : hashset -> t -> src:Bigstringaf.t -> src_off:int -> unit
+  val next : hashset -> t -> t
+  val iter_all : hashset -> f:(t -> unit) -> unit
+  val to_offset : t -> int
+end = struct
+  type t = Offset of int [@@ocaml.unboxed]
+
+  let offset_of_hash h hash =
+    let index = abs hash mod h.slot_count in
+    Offset (index * h.elt_length)
+
+  let of_elt h elt = offset_of_hash h (h.hash_elt elt)
+
+  let of_elt_substring h ~src ~src_off =
+    offset_of_hash h (h.hash_elt_substring src ~off:src_off ~len:h.elt_length)
+
+  let contains h (Offset offset) string =
+    Bigstringaf.memcmp_string h.data offset string 0 h.elt_length = 0
+
+  let contains_substring h (Offset offset) ~src ~src_off =
+    Bigstringaf.memcmp h.data offset src src_off h.elt_length = 0
+
+  let is_empty h t = contains h t h.empty_slot
+
+  let get h (Offset offset) =
+    Bigstringaf.substring h.data ~off:offset ~len:h.elt_length
+
+  let set h (Offset offset) elt =
+    Bigstringaf.blit_from_string elt ~src_off:0 h.data ~dst_off:offset
+      ~len:h.elt_length
+
+  let set_substring h (Offset offset) ~src ~src_off =
+    Bigstringaf.blit src ~src_off h.data ~dst_off:offset ~len:h.elt_length
+
+  let next h (Offset offset) = Offset ((offset + h.elt_length) mod h.data_length)
+
+  let iter_all hashset ~f =
+    assert (hashset.data_length <> 0);
+    f (Offset 0);
+    let rec aux = function
+      | Offset 0 -> ()
+      | offset ->
+          f offset;
+          aux (next hashset offset)
+    in
+    aux (next hashset (Offset 0))
+
+  let to_offset (Offset n) = n
+end
+
+let empty_all_slots t =
+  Slot.iter_all t ~f:(fun slot -> Slot.set t slot t.empty_slot)
+
+module Default = struct
+  let hash : string -> int = Hashtbl.hash
+  let hash_substring t ~off ~len = hash (Bigstringaf.substring t ~off ~len)
+  let null ~elt_length = String.make elt_length '\000'
+end
+
+let create ~elt_length ?(initial_capacity = 0) ?hash ?hash_substring ?null () =
+  if elt_length <= 0 then
+    Fmt.invalid_arg "%s.create: element length must be strictly positive"
+      __MODULE__;
+  let empty_slot =
+    match null with Some x -> x | None -> Default.null ~elt_length
+  in
+  let hash_elt, hash_elt_substring =
+    match (hash, hash_substring) with
+    | Some h, Some h' -> (h, h')
+    | None, None -> (Default.hash, Default.hash_substring)
+    | Some _, None | None, Some _ ->
+        Fmt.invalid_arg
+          "%s.create: must pass either both [hash] and [hash_substring] or \
+           neither"
+          __MODULE__
+  in
+  let slot_count =
+    let rec aux n =
+      if n >= initial_capacity then n
+      else if n * 2 > Sys.max_array_length then n
+      else aux (n * 2)
+    in
+    aux 2
```

Comment:
I suggest that `let slot_count = initial_capacity` and that this function raises an `Invalid_arg` exception if `initial_capacity` is not a power of 2

Reply:
I see.

I would rename `initial_capacity` to `initial_slots`, documenting in the docstring that `initial_slots` is not the initial capacity of the set.